fairscale/experimental/optim/dynamic_loss_scaler.py

```diff
@@ -0,0 +1,207 @@
+            if pct_overflow >= self.tolerance:
+                self._decrease_loss_scale()
+                self._last_rescale_iter = self._iter
+                self._overflows_since_rescale = 0
+
+            if self.loss_scale <= self.min_loss_scale:
+                # Use FloatingPointError as an uncommon error that parent
+                # functions can safely catch to stop training.
+                self.loss_scale = prev_scale
+                raise FloatingPointError(
+                    (
+                        "Minimum loss scale reached ({}). Your loss is probably exploding. "
+                        "Try lowering the learning rate, using gradient clipping or "
+                        "increasing the batch size."
+                    ).format(self.min_loss_scale)
+                )
+
+            self._iter += 1
+            raise OverflowError("setting loss scale to: " + str(self.loss_scale))
+
+    def update(self) -> None:
+        """
+        Updates the scale factor.
+        """
+
+        if (self._iter - self._last_overflow_iter) % self.scale_window == 0:
+            self.loss_scale *= self.scale_factor
+            self._last_rescale_iter = self._iter
+        self._iter += 1
+        self._per_optimizer_states = defaultdict(_refresh_per_optimizer_state)
+
+    def step(self, optimizer, *args, **kwargs):  # type: ignore
+        """
+        :meth:`step` unscale the gradients and step the optimizer.
+
+        ``*args`` and ``**kwargs`` are forwarded to ``optimizer.step()``.
+
+        Returns the return value of ``optimizer.step(*args, **kwargs)``.
+
+        Args:
+            optimizer (torch.optim.Optimizer):  Optimizer that applies the gradients.
+            args:  Any arguments.
+            kwargs:  Any keyword arguments.
+        """
+        if "closure" in kwargs:
+            raise RuntimeError("Closure use is not currently supported if DynamicLossScaler is enabled.")
+
+        optimizer_state = self._per_optimizer_states[id(optimizer)]
+
+        if optimizer_state is OptState.STEPPED:
+            raise RuntimeError("step() has already been called since the last update().")
+
+        # check gradient norm. If gradient norm is nan or inf, adjust scale here, and skip step.
+        # clip_grads_norm can happen before this step
+        for group in optimizer.param_groups:
+            grad_norm = self._get_gradients_norm(group["params"])
+            try:
+                self._check_overflow(grad_norm)
+            except OverflowError:
+                return None
+
+        if optimizer_state is OptState.READY:
+            self.unscale_(optimizer)
+            state_dict = optimizer.state_dict()
+            state_dict["loss_scale"] = self.loss_scale
+
+        retval = optimizer.step(*args, **kwargs)
+
+        optimizer_state = OptState.STEPPED
+
+        return retval
+
+    def unscale_(self, optimizer: torch.optim.Optimizer) -> None:
+
```

Comment:
missing doc string?

Reply:
This is internal method. Doc string is not necessary. I can add some comments here.

Reply:
In that case I would remove this line or add more details.